proposals/0245-sharing-wifi-ssid-and-password.md

```diff
@@ -18,627 +18,1205 @@ If WiFi credentials are automatically generated by the IVI system, the password
-                m_wifiManager.disconnect();
-                m_wifiManager.enableNetwork(i.networkId, true);
-                m_wifiManager.reconnect();
-                break;
+```json
+"policy_table": {
+    ...
+    "functional_groupings": {
+        ...
+        "NetworkSharing": {
+            "rpcs": {
+                "OnNetworkConfiguration": {
+                    "hmi_levels": [
+                        "BACKGROUND",
+                        "FULL",
+                        "LIMITED",
+                        "NONE"
+                    ]
+                },
+                "JoinNetwork": {
+                    "encryption_required": true,
+                    "hmi_levels": [
+                        "BACKGROUND",
+                        "FULL",
+                        "LIMITED",
+                        "NONE"
+                    ]
+                }
             }
         }
     }
-
-    private void connectToWiFi_API_29(String ssid,String password){
-		  WifiNetworkSpecifier suggestion2 =
-				  new WifiNetworkSpecifier.Builder()
-						.setSsid(ssid)
-						.setWpa2Passphrase(password)
-						.build();
-		  List<WifiNetworkSpecifier> suggestionsList =
-				  new ArrayList<>();
-		  suggestionsList.add(suggestion2);
-
-		  NetworkRequest networkRequest = new NetworkRequest.Builder()
-				.addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
-				.setNetworkSpecifier(suggestion2)
-				.build();
-
-		  m_conectivityManager.requestNetwork(networkRequest,networkCallback);
-    }
+}
 ```
-Permissions required to use above mentioned API's are as follows:
+
+__Note:__ Keep in mind, if `encryption_required` for `OnNetworkConfiguration` is set to `true`, Mobile application will not receive notification until secure service is not established.
+
+## Troubleshooting and race conditions
+
+### Race condition in case of multiple device connection
+
+#### Registration race
+
+From the Core perspective registration process is sequential and since it will cache registration order, there gonna be no race condition during registration.
+
+#### Capability race
+
+##### Vehicle is the hotspot
+
+After the application from the device is connected and Vehicle should be hotspot, Core will send `OnNetworkConfiguration` notification to the application and to HMI. Since it is Vehicle responsibility to start Hotspot, Core should wait for `OnCapabilityUpdate` with the updated capabilities. Only after this notification, Core is ready to process feature enabled application from another device.
+
+#### Mobile device is the hotspot
+
+After Core has decided, that Mobile device should be an Access point, Core will send `OnNetworkConfiguration` notification to the application and HMI. Since HMI has nothing to do before establishing WiFi connection. So the HMI capability will be updated only after connection attempt to the Mobile hotspot. Before this point, Core will not be able to process feature enabled applications from other devices.

```

Comment:
```suggestion
After SDL Core has decided that the mobile device should be an access point, Core will send `OnNetworkConfiguration` notification to the application and HMI. Since HMI has nothing to do before establishing WiFi connection, the HMI capability will be updated only after connection attempt to the mobile hotspot. Before this point, Core will not be able to process feature enabled applications from other devices.
```